### PR TITLE
Bump sonata-project/datagrid-bundle to version ^3.0

### PR DESCRIPTION
**Pull request by wbloszyk**
## Subject

I am targeting this branch, because we should bump in this branch. It is potential BC-break.

## Changelog

```markdown
### Added
- support for sonata-project/datagrid-bundle to version ^3.0
### Removed
- support for sonata-project/datagrid-bundle to version < 3.0
```

### CODE REVIEW

File: UPGRADE-4.x.md [hunk 1]
```diff
@@ -1,6 +1,15 @@
 UPGRADE 4.x
 ===========
 
+UPGRADE FROM 3.x to 3.x
```
Comment (VincentLanglet) [inline UPGRADE-4.x.md:4]: UPGRADE FROM 4.x to 4.x

Comment (wbloszyk) [reply UPGRADE-4.x.md:4]: Copy/Paste bug :smile:

Comment (wbloszyk) [reply UPGRADE-4.x.md]: Fixed.

**Review by VincentLanglet — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Test are failing

**Comment by wbloszyk** [comment]
@VincentLanglet Test are failing becouse API. The will be fix in #1225.

File: UPGRADE-4.x.md [hunk 2]
```diff
@@ -1,9 +1,15 @@
 UPGRADE 4.x
 ===========
 
-UPGRADE FROM 3.x to 3.x
+UPGRADE FROM 4.x to 4.x
 =======================
 
+### Upgrade to SonataDatagridBundle 3.0
+
+If you are extending these method you should add parameter and return type hints:
```
Comment (phansys) [inline UPGRADE-4.x.md:9]:
```suggestion
If you are extending these method you MUST add argument and return type declarations:
```

File: UPGRADE-4.x.md [hunk 3]
```diff
@@ -1,9 +1,15 @@
...
+- `Sonata\UserBundle\Entity\UserManager::getPager`
```
Comment (phansys) [inline UPGRADE-4.x.md:10]:
```suggestion
- `Sonata\UserBundle\Entity\UserManager::getPager()`
```

File: UPGRADE-4.x.md [hunk 4]
```diff
@@ -1,9 +1,15 @@
...
+- `Sonata\UserBundle\Entity\GroupManager::getPager`
```
Comment (phansys) [inline UPGRADE-4.x.md:11]:
```suggestion
- `Sonata\UserBundle\Entity\GroupManager::getPager()`
```

File: tests/Controller/Api/UserControllerTest.php [hunk 5]
```diff
@@ -35,19 +36,22 @@ class UserControllerTest extends TestCase
 {
     public function testGetUsersAction(): void
     {
+        $user = $this->createMock(UserInterface::class);
         $userManager = $this->createMock(UserManagerInterface::class);
-        $userManager->expects($this->once())->method('getPager')->willReturn([]);
+        $pager = $this->createMock(PagerInterface::class);
```
Comment (phansys) [inline tests/Controller/Api/UserControllerTest.php:41]:
```suggestion
        $pager = $this->createStub(PagerInterface::class);
```

File: tests/Controller/Api/UserControllerTest.php [hunk 6]
```diff
@@ -35,19 +36,22 @@ class UserControllerTest extends TestCase
...
         $paramFetcher->expects($this->once())->method('all')->willReturn([]);
 
-        $this->assertSame([], $this->createUserController(null, $userManager)->getUsersAction($paramFetcher));
+        $this->assertSame([$user], $this->createUserController(null, $userManager)->getUsersAction($paramFetcher)->getResults());
     }
 
     public function testGetUserAction(): void
     {
-        $user = $this->createMock(\Sonata\UserBundle\Model\UserInterface::class);
+        $user = $this->createMock(UserInterface::class);
```
Comment (phansys) [inline tests/Controller/Api/UserControllerTest.php:54]:
```suggestion
        $user = $this->createStub(UserInterface::class);
```

File: tests/Controller/Api/GroupControllerTest.php [hunk 7]
```diff
@@ -35,14 +36,16 @@ public function testGetGroupsAction(): void
     {
         $group = $this->createMock(GroupInterface::class);
         $groupManager = $this->createMock(GroupManagerInterface::class);
-        $groupManager->expects($this->once())->method('getPager')->willReturn([$group]);
+        $pager = $this->createMock(PagerInterface::class);
```
Comment (phansys) [inline tests/Controller/Api/GroupControllerTest.php:39]:
```suggestion
        $pager = $this->createStub(PagerInterface::class);
```

**Review by phansys — CHANGES_REQUESTED** [review]

File: UPGRADE-4.x.md [hunk 8]
```diff
@@ -1,9 +1,15 @@
 UPGRADE 4.x
 ===========
 
-UPGRADE FROM 3.x to 3.x
+UPGRADE FROM 4.x to 4.x
 =======================
 
+### Upgrade to SonataDatagridBundle 3.0
+
```
Comment (phansys) [inline UPGRADE-4.x.md:8]: With this approach, we should later group also the required changes for "nelmio/api-doc-bundle" in this section.
IMO, we should stop using bundle aliases (like "SonataDatagridBundle") in order to be specific about the package names.
```suggestion
### Dependencies

 - "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.
```

File: UPGRADE-4.x.md [hunk 9]
```diff
@@ -4,8 +4,9 @@ UPGRADE 4.x
 UPGRADE FROM 4.x to 4.x
 =======================
 
-### Upgrade to SonataDatagridBundle 3.0
+### Dependencies
 
+ - "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.
```
Comment (phansys) [inline UPGRADE-4.x.md:9]: Just to give more context to my previous suggestion about this change. I'm proposing something like:
```md
### Dependencies

- "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.
  Some description about the upgrade process for this dependency
  with all the required information.

- "package/a" is bumped from ^1.0 to ^2.0.
  Some description about the upgrade process for this dependency
  with all the required information.
```

File: UPGRADE-4.x.md [hunk 10]
```diff
@@ -1,12 +1,19 @@
...
-### Support for NelmioApiDocBundle > 3.6 is added
+### Dependencies
 
-Controllers for NelmioApiDocBundle v2 were moved under `Sonata\UserBundle\Controller\Api\Legacy\` namespace and controllers for NelmioApiDocBundle v3 were added as replacement. If you extend them, you must ensure they are using the corresponding inheritance.
+- "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.
+If you are extending these method you MUST add argument and return type declarations:
+  - `Sonata\UserBundle\Entity\UserManager::getPager()`
+  - `Sonata\UserBundle\Entity\GroupManager::getPager()`
+
+- "nelmio/api-doc-bundle" from now is also supported for ^3.6
```
Comment (phansys) [inline UPGRADE-4.x.md:14]: The term "now" is relative in a document like this. IMO, we should avoid to make use of these kind of references. The relevant concept here is the version number.
```suggestion
- Added support for "nelmio/api-doc-bundle" ^3.6
```

File: UPGRADE-4.x.md [hunk 11]
```diff
@@ -1,12 +1,19 @@
...
 
-UPGRADE FROM 3.x to 3.x
+UPGRADE FROM 4.x to 4.x
 =======================
 
-### Support for NelmioApiDocBundle > 3.6 is added
+### Dependencies
 
-Controllers for NelmioApiDocBundle v2 were moved under `Sonata\UserBundle\Controller\Api\Legacy\` namespace and controllers for NelmioApiDocBundle v3 were added as replacement. If you extend them, you must ensure they are using the corresponding inheritance.
+- "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.
```
Comment (phansys) [inline UPGRADE-4.x.md:9]: image
```suggestion
- "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.

```

File: UPGRADE-4.x.md [hunk 12]
```diff
@@ -1,12 +1,19 @@
...
+If you are extending these method you MUST add argument and return type declarations:
```
Comment (phansys) [inline UPGRADE-4.x.md:10]:
```suggestion
  If you are extending these method you MUST add argument and return type declarations:
```

**Review by phansys — CHANGES_REQUESTED** [review]

**Review by phansys — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Comment by wbloszyk** [comment]
@phansys Can you re-check it?

File: UPGRADE-4.x.md [hunk 13]
```diff
@@ -1,12 +1,20 @@
...
+UPGRADE FROM 4.x to 4.x
 =======================
 
-### Support for NelmioApiDocBundle > 3.6 is added
+### Dependencies
 
-Controllers for NelmioApiDocBundle v2 were moved under `Sonata\UserBundle\Controller\Api\Legacy\` namespace and controllers for NelmioApiDocBundle v3 were added as replacement. If you extend them, you must ensure they are using the corresponding inheritance.
+- "sonata-project/datagrid-bundle" is bumped from ^2.4 to ^3.0.
+
+  If you are extending these method you MUST add argument and return type declarations:
```
Comment (phansys) [inline UPGRADE-4.x.md:11]:
```suggestion
  If you are extending these methods, you MUST add argument and return type declarations:
```

**Review by phansys — APPROVED** [review]

**Review by VincentLanglet — APPROVED** [review]

**Review by phansys — APPROVED** [review]

**Comment by VincentLanglet** [comment]
Thanks

**Pull request merged**